### PR TITLE
Detect newer available version

### LINKED ISSUE #109: Detect newer available version
It could be suggested in wrapper generator. There are two cases:
* we generate wrappers for versions e.g. v2.0.1, v2.1 (creates wrapper `V2`), and there's a new version v2.2 - then it needs bumping the version in the existing wrapper request
* we generate wrappers for major versions e.g. v2, v3, and there's a new version v4 - then suggest adding a new wrapper request

### LINKED ISSUE #119: [Core feature request] concurrency
**What feature do you need?**

- https://docs.github.com/en/actions/using-jobs/using-concurrency
- https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idconcurrency

**Do you have an example usage?**

```yaml
concurrency: staging_environment
```

```yaml
concurrency: 
  group: ${{ github.ref }}
  cancel-in-progress: true
```

Keep in mind that it's available both
- at a workflow level
- at `jobs.<job_id>.concurrency` level

### PR DESCRIPTION
**Pull request by jmfayard**
resolves #109 

see result here: https://github.com/krzema12/github-actions-kotlin-dsl/issues/171

### CODE REVIEW

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.data.Table2
+import io.kotest.data.forAll
+import io.kotest.data.headers
+import io.kotest.data.row
+import io.kotest.data.table
+import io.kotest.inspectors.forAll
+import io.kotest.matchers.shouldBe
+
+class SuggestVersionsKtTest : FunSpec({
```
Comment (krzema12) [inline wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt:13]:
```suggestion
class SuggestVersionsTest : FunSpec({
```

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    val githubAuthorization = System.getenv("GITHUB_TOKEN")
+        ?: error(
+            """
+            |Missing environment variable export GITHUB_TOKEN=token
+            |Create a personal token at https://github.com/settings/tokens
+           """.trimMargin()
+        )
+
+    val actionsMap = wrappersToGenerate
+        // .take(3)
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:33]:
```suggestion
```

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+@kotlinx.serialization.Serializable
+data class GithubTag(
+    val ref: String,
+)
+
+val json = Json { ignoreUnknownKeys = true }
+
+val okhttpClient by lazy {
+    OkHttpClient()
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:20]: A matter of taste, but how about using Ktor client? :) Supports coroutines and (de)serialization with kotlinx.serialization out of the box, is more Kotlin-centric, and I'm personally more familiar with it. I generally follow a rule that there's a library from JetBrains for a certain task, I use it in the first order, and fall back to third-party only if necessary.

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+fun ActionCoords.fetchAvailableVersions(GITHUB_TOKEN: String): List<String> {
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:23]:
```suggestion
fun ActionCoords.fetchAvailableVersions(githubToken: String): List<String> {
```

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        .map { it.actionCoords }
+        .groupBy { "${it.owner}/${it.name}" }
+    for (list in actionsMap.values) {
+        val coords = list.maxByOrNull { it.version }!!
+        val existingVersions = list.map { it.version }
+        val available = coords.fetchAvailableVersions(githubAuthorization)
+        suggestNewerVersion(existingVersions, available)?.let { message ->
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:40]: Could we make it return non-zero if some suggestion appears? Thanks to this we'll be able to run it automatically through GitHub workflow and if it exits with non-zero code, it will make the workflow fail and I'll get an e-mail ;) If you have a better idea for the API for this command that would fit this scenario, I'm open.

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            |Create a personal token at https://github.com/settings/tokens
+           """.trimMargin()
+        )
+
+    val actionsMap = wrappersToGenerate
+        // .take(3)
+        .map { it.actionCoords }
+        .groupBy { "${it.owner}/${it.name}" }
+    for (list in actionsMap.values) {
+        val coords = list.maxByOrNull { it.version }!!
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:37]: I try to avoid `!!` to not get mysterious NPEs. Let's use e.g. `?: error("Could not find newest version")` or something similar.

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ * new major version(s) available: [v3] for ActionCoords("actions", "download-artifact", "v2")
+ * new minor version available: v2.2.0 for ActionCoords("peterjgrainger", "action-create-branch", "v2.1.0")
+ * ```
+ */
+fun main() {
+    val githubAuthorization = System.getenv("GITHUB_TOKEN")
+        ?: error(
+            """
+            |Missing environment variable export GITHUB_TOKEN=token
+            |Create a personal token at https://github.com/settings/tokens
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:28]: Let's write that the token needs to have `public_repo` scope.

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            "v1.0.0",
+            "v1.0.9",
+            "v1.0.12",
+            "v1.2.0",
+            "v1.9.0",
+            "v1.13.1",
+            "v9.0.0",
+            "v12.0.0",
+        )
+        sortedVersions.sortedBy { VersionComparable(it) } shouldBe sortedVersions
```
Comment (krzema12) [inline wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt:25]: Wait, this test tests almost nothing :) You give it already sorted list.

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            testCases.forAll { current, available ->
+                suggestNewerVersion(current.split(", "), available.split(", ")) shouldBe null
+            }
+        }
+
+        test("New major version") {
+            assertSoftly {
+
+                suggestNewerVersion(
+                    listOf("v1"), listOf("v1.1.1", "v2.0.1", "v3.1.1")
```
Comment (krzema12) [inline wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt:50]: If these tests are for major versions, without minor and patch, I think it should say e.g.
```suggestion
                    listOf("v1"), listOf("v1", "v2", "v3")
```
Same below.

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+            assertSoftly {
+                suggestNewerVersion(
+                    listOf("v2.1.0", "v1.0.4"),
+                    listOf("v2.1.0", "v1.0.4", "v2.3.0", "v1.0.5"),
+                ) shouldBe "new minor version available: v2.3.0"
+
+                suggestNewerVersion(
+                    listOf("v9.0.1"),
+                    listOf("v12.1.2"),
+                ) shouldBe "new minor version available: v12.1.2"
```
Comment (krzema12) [inline wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt:89]: The message here is misleading, actually the new major version is available.

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+fun suggestNewerVersion(existingVersions: List<String>, availableVersions: List<String>): String? {
+    val maxExisting = existingVersions.maxByOrNull { VersionComparable(it) }!!
+
+    if (existingVersions.all { isMajorVersion(it) }) {
+        val majorVersions = availableVersions
+            .map { it.substringBefore(".") }
+            .distinct()
+            .sortedBy { VersionComparable(it) }
+        val newer = majorVersions.filter { VersionComparable(it) > VersionComparable(maxExisting) }
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:57]: Instead of keeping versions as strings and wrapping them in `VersionComparable` wherever needed, I suggest renaming `VersionComparable` to just `Version`, instantiating it from strings as soon as possible (e.g. `maxExisting` could already be an instance of `Version`), and then intuitively using it like `it > maxExisting` here.

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+import it.krzeminski.githubactions.wrappergenerator.domain.ActionCoords
+import kotlinx.serialization.builtins.ListSerializer
+import kotlinx.serialization.json.Json
+import okhttp3.OkHttpClient
+import okhttp3.Request
+
+val ActionCoords.apiTagsUrl: String
+    get() = "https://api.github.com/repos/$owner/$name/git/matching-refs/tags/v"
+
+@kotlinx.serialization.Serializable
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:12]:
```suggestion
@Serializable
```

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+fun ActionCoords.fetchAvailableVersions(GITHUB_TOKEN: String): List<String> {
+    val request: Request = Request.Builder()
+        .header("Authorization", "token $GITHUB_TOKEN")
+        .url(apiTagsUrl)
+        .build()
+
+    val content = okhttpClient.newCall(request).execute().use { response ->
+        if (response.isSuccessful.not()) {
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:30]: Is there any advantage of this notation over
```suggestion
        if (!response.isSuccessful) {
```
? Is `not()` more prominent in your opinion?

**Review by krzema12 — CHANGES_REQUESTED** [review]

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt (hunk 5 above)
Comment (jmfayard) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:40]: I was thinking that if the output is not blank, we open an issue with its content exactly like https://github.com/krzema12/github-actions-kotlin-dsl/issues/171

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt (hunk 8 above)
Comment (jmfayard) [reply wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt:25]: Well I assume that `sortedBy { }` is already battle tested and I'm testing `VersionComparable.compareTo()`.
To make things clearer, I have added `Iterable.shuffled()` which change the order

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt (hunk 3 above)
Comment (jmfayard) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:20]: @krzema12 I thought about ktor vs okhttp, and for me the main question is: 
- do we want to use coroutines to make things faster by fetching in parallel ?
- or do we want to continue with blocking calls to keep things simpler ?

If we want to use coroutines, we should probably do it in the wrapper generators too (instead of `URI(..).readText()`) because that's the one we run most often.

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt (hunk 9 above)
Comment (jmfayard) [reply wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt:50]: That's not how it works.
The list `listOf("v1")` comes from the versions defined in `wrappersToGenerate`.
The list `listOf("v1.1.1", "v2.0.1", "v3.1.1")` is what comes from the HTTP call, and from that we have to understand that `v2` and `v3`  are available.

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt (hunk 10 above)
Comment (jmfayard) [reply wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt:89]: What would you suggest ? 
`new hardcoded version` ?

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt (hunk 13 above)
Comment (jmfayard) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:30]: I prefer the `not()` because it keeps a coherent left-to-right reading order.

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt (hunk 11 above)
Comment (jmfayard) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:57]: Wow that's a great sugestion thanks

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt (hunk 3 above)
Comment (krzema12) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:21]: Right, fetching wrapper metadata in parallel would be a nice thing - we can think about it in the future.
Let's leave okhttp for now.

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt (hunk 5 above)
Comment (krzema12) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:41]: Ok, got it - we can try with such API, it's anyway internal to the project so not a big deal. Good idea with automatic issue creation! Are you planning to add such workflow?

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt (hunk 10 above)
Comment (krzema12) [reply wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt]: Maybe simply `new version available`?

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+            println(response.headers)
+            error("API rate reached?  See https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting")
+        }
+        response.body!!.string()
+    }
+    val data = json.decodeFromString(ListSerializer(GithubTag.serializer()), content)
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:37]: I think this simpler form will work as well:
```suggestion
    val data = json.decodeFromString<List<GithubTag>>(content)
```

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsTest.kt [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+        test("New major version") {
+            assertSoftly {
+
+                suggestNewerVersion(
+                    "v1".versions(), "v1.1.1, v2.0.1, v3.1.1".versions()
+                ) shouldBe "new major version(s) available: [v2, v3]"
+
+                suggestNewerVersion(
+                    "v2".versions(), "v1.1.1, v2.0.1, v3.1.1".versions()
```
Comment (krzema12) [inline wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsTest.kt:60]: Continuation of https://github.com/krzema12/github-actions-kotlin-dsl/pull/170#discussion_r846739735

I get what you meant, but suggesting that `v3` is available where there's no explicit `v3` tag for a given action looks like an overinterpretation. I think we should suggest `v3` only if the tag `v3` exists, and if a tag with minor/path exists, we shouldn't suggest it  - it would then mean that the action maintainer messed something with released versions and didn't release the major version-based one.

**Review by krzema12 — CHANGES_REQUESTED** [review]

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt (hunk 14 above)
Comment (jmfayard) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:37]: No, that doesn't work because of type erasure.

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsTest.kt (hunk 15 above)
Comment (jmfayard) [reply wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsTest.kt]: You were right, I thought the tag "v2" weren't available in the APi results but they actually are.
I have fixed the logic.
I also now detect if we are using an hardcoded version like `v2.1.0` but an explicit version `v2` is available

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt (hunk 9 above)
Comment (jmfayard) [reply wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsKtTest.kt]: I was wrong, logic updated, see comment below

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt (hunk 5 above)
Comment (jmfayard) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:41]: Yes, almost written already, will submit in a separate pull request

**Comment by jmfayard** [comment]
@krzema12 I am almost done, except for the workflow to create an issue, which I will submit in a separate pull request

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt (hunk 14 above)
Comment (krzema12) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:37]: "Works on my machine" ;) This overload of `decodeFromString` is `reified`:

```kotlin
fun main() {
    val content = "[{\"ref\": \"some-ref\"}, {\"ref\": \"12334\"}]"
    val data = json.decodeFromString<List<GithubTag>>(content)
    println(data)  // [GithubTag(ref=some-ref), GithubTag(ref=12334)]
}
```

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+ * You need to set an environment variable
+ *  `$ export GITHUB_TOKEN=token`
+ *
+ *  Create your token at https://github.com/settings/tokens
+ *
+ *  Usage:
+ *
+ *  ```
+ * $ ./gradlew :wrapper-generator:suggestVersions
+ * new major version(s) available: [v3] for ActionCoords("actions", "cache", "v2")
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:19]: I think this output format is outdated, it now displays either `""` or `"*"` as version. It looks a bit weird. Could we either go back to what this comment shows, or if there's some problem with it, update this comment?

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Version.kt [hunk 17]
```diff
@@ -0,0 +1,26 @@
+package it.krzeminski.githubactions.wrappergenerator.versions
+
+data class Version(val version: String) : Comparable<Version> {
+    val input: String = version.removePrefix("v").removePrefix("V")
+    val major = input.substringBefore(".").toIntOrNull() ?: 0
+    val minor = input.substringAfter(".").substringBefore(".").toIntOrNull() ?: 0
+    val patch = input.substringAfterLast(".").toIntOrNull() ?: 0
+    override fun compareTo(other: Version): Int {
```
Comment (krzema12) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Version.kt:8]: nit. some spacing before the function please:
```suggestion

    override fun compareTo(other: Version): Int {
```

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsTest.kt [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.data.Table2
+import io.kotest.data.forAll
+import io.kotest.data.headers
+import io.kotest.data.row
+import io.kotest.data.table
+import io.kotest.inspectors.forAll
+import io.kotest.matchers.shouldBe
+
+class SuggestVersionsTest : FunSpec({
```
Comment (krzema12) [inline wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsTest.kt:13]: Let's have a test for this peculiar edge case of some actions that specify the version as `latest`, or to generify, something that doesn't resemble a version number:

https://github.com/krzema12/github-actions-kotlin-dsl/blob/a7b9bb4a98e9ce1ac79cf00081f9f0ae998f08b9/wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/WrappersToGenerate.kt#L588

**Review by krzema12 — CHANGES_REQUESTED** [review]

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt (hunk 16 above)
Comment (jmfayard) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:19]: I didn't know what to put there since we coulud have mulutiple versions

Comment (krzema12) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt:19]: Ok, so updating the comment or maybe even removing it (so that it won't get outdated anymore) will be ok :)

Comment (jmfayard) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersions.kt]: done

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Version.kt (hunk 17 above)
Comment (jmfayard) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Version.kt:9]: Done

File: wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsTest.kt (hunk 18 above)
Comment (jmfayard) [reply wrapper-generator/src/test/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/SuggestVersionsTest.kt:13]: done

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+        .build()
+
+    val content = okhttpClient.newCall(request).execute().use { response ->
+        if (response.isSuccessful.not()) {
+            println(response.headers)
+            error("API rate reached?  See https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting")
+        }
+        response.body!!.string()
+    }
+    val data = json.decodeFromString<List<GithubTag>>(content)
```
Comment (jmfayard) [inline wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt:37]: You were right, I didn't have the right import

**Comment by jmfayard** [comment]
I'm done here

File: wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt (hunk 14 above)
Comment (jmfayard) [reply wrapper-generator/src/main/kotlin/it/krzeminski/githubactions/wrappergenerator/versions/Http.kt]: You were right, fixed
